lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,269 @@
+    """
+    def create_from_blockrecords(self, blocks):
+        """
+        Creates a new VisibleBlocks model object.
+
+        Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        if not isinstance(blocks, BlockRecordList):
+            blocks = BlockRecordList(blocks)
+        model, _ = self.get_or_create(hashed=blocks.to_hash(), defaults={'blocks_json': blocks.to_json()})
+        return model
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a
+    given subsection at the time they are used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored
+    in the blocks_json field. A hash of this json array is used for lookup
+    purposes.
+    """
+    blocks_json = models.TextField()
+    hashed = models.CharField(max_length=44, unique=True)
+
+    objects = VisibleBlocksQuerySet.as_manager()
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return "VisibleBlocks object - hash:{}, raw json:'{}'".format(self.hashed, self.blocks_json)
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as a list of
+        BlockRecords in the order they were provided.
+        """
+        block_dicts = json.loads(self.blocks_json)
+        record_generator = (BlockRecord(data['locator'], data['weight'], data['max_score']) for data in block_dicts)
+        return BlockRecordList(record_generator)
+
+
+class PersistentSubsectionGradeQuerySet(models.QuerySet):
+    """
+    A custom QuerySet, that handles creating a VisibleBlocks model on creation, and
+    extracts the course id from the provided usage_key.
+    """
+    def create(self, **kwargs):
+        """
+        Instantiates a new model instance after creating a VisibleBlocks instance.
+
+        Arguments:
+            user_id (int)
+            usage_key (serialized UsageKey)
+            course_version (str)
+            subtree_edited_date (datetime)
+            earned_all (float)
+            possible_all (float)
+            earned_graded (float)
+            possible_graded (float)
+            visible_blocks (iterable of BlockRecord)
+        """
+        visible_blocks = kwargs.pop('visible_blocks')
+        if not isinstance(visible_blocks, BlockRecordList):
+            visible_blocks = BlockRecordList(visible_blocks)
+        try:
+            visible_blocks_model = VisibleBlocks.objects.create_from_blockrecords(blocks=visible_blocks)
+        except IntegrityError:
+            # If an integrity error occurs, the VisibleBlocks model we want to
+            # create already exists.  Use the hash for our current list of
+            # blocks as the foreign key.
+            visible_blocks_hash = visible_blocks.to_hash()
+        else:
+            visible_blocks_hash = visible_blocks_model.hashed
```

Comment:
I've been thinking about it.  My concern is that moving it into `create_from_blockrecords` makes an irritatingly oblique signature for a QuerySet constructor.  The creation of the actual `visible_blocks` object is tucked away as a side-effect, and the returned value is a hash.  It pushes the IO too far down the stack for this abstraction level.

Still trying to come up with a clean architecture for this.